aten/src/ATen/native/sparse/SparseBinaryOpIntersectionCommon.h

```diff
@@ -427,175 +402,25 @@ void _sparse_binary_op_intersection_kernel_impl(
-  // We would like to respect order in value intersection.
-  auto [lhs, lhs_selected, rhs, rhs_selected] = [&]() -> auto {
-    // Either source <=> x, ...
-    if (source.is_same(x)) {
-      return std::make_tuple(source, selected_source, probably_coalesced, selected_probably_coalesced);
-    // ... or source <=> y.
-    } else {
-      return std::make_tuple(probably_coalesced, selected_probably_coalesced, source, selected_source);
-    }
-  }();
-  auto res_values = value_selection_intersection_kernel_t::apply(
-      lhs._values().to(binary_op_res_dtype), // promote for better accuracy
-      lhs_selected,
-      rhs._values().to(binary_op_res_dtype), // promote for better accuracy
-      rhs_selected);
-  // Convert back if the promoted dtype is different from res.dtype.
-  // This could happen for in-place usage cases.
-  res_values = res_values.to(res.scalar_type());
-
+  const auto res_indices = source._indices().clone();
+  const auto binary_op_res_dtype = at::result_type(source._values(), probably_coalesced._values());
+  const auto res_values = value_selection_intersection_kernel_t::apply(
+      source._values().to(binary_op_res_dtype),
+      nnz_arange.narrow(-1, 0, source._nnz()),
+      probably_coalesced._values().to(binary_op_res_dtype),
+      intersection_first_idx.to(nnz_arange.scalar_type()),
+      intersection_count,
+      argsort_hash).to(res.scalar_type());
   const auto res_sparse_dim = source.sparse_dim();
-  const auto res_dense_dim = res_values.dim() - 1;
+  const auto res_dense_dim = source.dense_dim();
   const auto& res_shape = broadcasted_shape;
-  const auto res_nnz = selected_source.numel();
+  const auto res_nnz = source._nnz();
 
   auto* res_sparse_impl = get_sparse_impl(res);
   res_sparse_impl->raw_resize_(res_sparse_dim, res_dense_dim, res_shape);
   res_sparse_impl->set_indices_and_values_unsafe(res_indices, res_values);
   res_sparse_impl->set_nnz_and_narrow(res_nnz);
-  // Result is coalesced iff arguments are coalesced, conditioned on the fact
-  // that we do not check that intersection hash values are sorted and unique.
-  // <= : intersection contains only unique indices (or empty), and the algorithm's
-  // behavior is order-preserving. So, the result has only unique indices (or empty) which are sorted.
-  // => : proof by contraposition. The contrapositive statement reads
-  // `there is an uncoalesced argument => result is not coalesced`.
-  // If both arguments are uncoalesced, the result is clearly uncoalesced again
-  // thanks to the order-preserving behavior of the algorithm.
-  // Otherwise we have a coalesced argument `probably_coalesced` and an uncoalesced `source`.
-  // Since the matching beahavior of the algorithm respects the order of `source`, the result
-  // will be as coalesced as `source` is, which is uncoalesced.
-  res._coalesced_(source.is_coalesced() && probably_coalesced.is_coalesced());
+  res._coalesced_(y_.is_coalesced() || !distributive_with_sum);
```

Comment:
With `distributive_with_sum==true`, `x_.is_coalesced()==false`, `y_.is_coalesced()==true`, and `source` is the same as `x`, what guarantees that `res` is coalesced?

Reply:
It should be as coalesced as the source, I will fix that.